demos/sample_azure_iot/sample_azure_iot.c

```diff
@@ -0,0 +1,625 @@
+        TLS_Socket_Disconnect( &xNetworkContext );
+
+        *ppucIothubHostname = ucSampleIotHubHostname;
+        *pulIothubHostnameLength = ucSamplepIothubHostnameLength;
+        *ppucIothubDeviceId = ucSampleIotHubDeviceId;
+        *pulIothubDeviceIdLength = ucSamplepIothubDeviceIdLength;
+
+        return 0;
+    }
+
+#endif // democonfigENABLE_DPS_SAMPLE
+/*-----------------------------------------------------------*/
+
+/**
+* @brief Connect to server with backoff retries.
+*/
+static TlsTransportStatus_t prvConnectToServerWithBackoffRetries( const char * pcHostName,
+                                                                  uint32_t port,
+                                                                  NetworkCredentials_t * pxNetworkCredentials,
+                                                                  NetworkContext_t * pxNetworkContext )
+{
+    TlsTransportStatus_t xNetworkStatus;
+    BackoffAlgorithmStatus_t xBackoffAlgStatus = BackoffAlgorithmSuccess;
+    BackoffAlgorithmContext_t xReconnectParams;
+    uint16_t usNextRetryBackOff = 0U;
+
+    /* Initialize reconnect attempts and interval. */
+    BackoffAlgorithm_InitializeParams( &xReconnectParams,
+                                       sampleazureiotRETRY_BACKOFF_BASE_MS,
+                                       sampleazureiotRETRY_MAX_BACKOFF_DELAY_MS,
+                                       sampleazureiotRETRY_MAX_ATTEMPTS );
+
+    /* Attempt to connect to MQTT broker. If connection fails, retry after
+     * a timeout. Timeout value will exponentially increase till maximum
+     * attempts are reached.
+     */
+    do
+    {
+        LogInfo( ( "Creating a TLS connection to %s:%u.\r\n", pcHostName, port ) );
+        /* Attempt to create a mutually authenticated TLS connection. */
+        xNetworkStatus = TLS_Socket_Connect( pxNetworkContext,
+                                             pcHostName, port,
+                                             pxNetworkCredentials,
+                                             sampleazureiotTRANSPORT_SEND_RECV_TIMEOUT_MS,
+                                             sampleazureiotTRANSPORT_SEND_RECV_TIMEOUT_MS );
+
+        if( xNetworkStatus != TLS_TRANSPORT_SUCCESS )
+        {
+            /* Generate a random number and calculate backoff value (in milliseconds) for
+             * the next connection retry.
+             * Note: It is recommended to seed the random number generator with a device-specific
+             * entropy source so that possibility of multiple devices retrying failed network operations
+             * at similar intervals can be avoided. */
+            xBackoffAlgStatus = BackoffAlgorithm_GetNextBackoff( &xReconnectParams, configRAND32(), &usNextRetryBackOff );
+
+            if( xBackoffAlgStatus == BackoffAlgorithmRetriesExhausted )
+            {
+                LogError( ( "Connection to the broker failed, all attempts exhausted." ) );
```

Comment:
Please log the `xNetworkStatus` both here and even if we have additional retries left.  This is very helpful for troubleshooting.

Also when we're in the "Retrying..." log please print how many milliseconds it'll be till next retry.